src/algebra/smul_with_zero.lean

```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import group_theory.group_action.defs
+import algebra.algebra.basic
+/-!
+# Introduce `smul_with_zero`
+
+Alongside with a usual monoid action on a Type `M`, we introduce an action of a `monoid_with_zero`
+on a Type with `0`.
+
+In particular, for Types `R` and `M`, both containing `0`, we define `smul_with_zero R M` to
+be the typeclass where the products `r • 0` and `0 • m` vanish for all `r ∈ R` and all `m ∈ M`.
+
+Moreover, in the case in which `R` is a `monoid_with_zero`, we introduce the typeclass
+`mul_action_with_zero R M`, mimicking group actions and having an absorbing `0` in `R`.
+Thus, the action is required to be compatible with
+
+* the unit of the monoid, acting as the identity;
+* the zero of the monoid_with_zero, acting as zero;
+* associativity of the monoid.
+-/
+
+variables (R M : Type*)
+
+section has_zero
+
+variable [has_zero M]
+
+/--  `smul_with_zero` is a class consisting of a Type `R` with `0 : R` and a scalar multiplication
+of `R` on a Type `M` with `0`, such that the equality `r • m = 0` holds if at least one among `r`
+or `m` equals `0`. -/
+class smul_with_zero [has_zero R] extends has_scalar R M :=
+( smul_zero : ∀ r : R, r • (0 : M) = 0 )
+( zero_smul : ∀ m : M, (0 : R) • m = 0 )
+
+--attribute [simp] smul_with_zero.smul_zero smul_with_zero.zero_smul
+
+--namespace smul_with_zero
+
+/-
+-- todo: this should replace the global one
+lemma smul_zero [has_zero R] [smul_with_zero R M] (r : R) :
+  r • (0 : M) = 0 :=
+smul_zero' r
```

Comment:
After doing what I suggest in `module/basic`:
```suggestion
@[simp] lemma smul_zero [has_zero R] [smul_with_zero R M] (r : R) :
  r • (0 : M) = 0 :=
smul_with_zero .smul_zero' r
/-
```

Reply:
The problem with this is that Lean says that `smul_zero` already exists.

Reply:
Oh sorry, this one was meant to be `zero_smul` not `smul_zero`

Reply:
https://github.com/leanprover-community/mathlib/pull/6590/files#r593110232

Reply:
Ok, while I do not really understand what is going on, I think that I managed to fix all the broken proof in `basic`.  Somewhat disturbingly, I had to add `R M` as explicit arguments to a `smul_...`: I wonder whether this will mean a similar breakage in different files, though.